### PR TITLE
planner: support drop binding by sql digest

### PR DESCRIPTION
**Pull request by fzzf678**
### What problem does this PR solve?

Issue Number: ref #39199 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -373,6 +373,7 @@ import (
 	deallocate            "DEALLOCATE"
 	definer               "DEFINER"
 	delayKeyWrite         "DELAY_KEY_WRITE"
+	digest 				  "DIGEST"
```
Comment (qw4990) [inline parser/parser.y:376]: not formatted

**Review by qw4990 — COMMENTED** [review]
`parser_golden.y` should be removed?

File: executor/bind.go [hunk 2]
```diff
@@ -83,6 +88,17 @@ func (e *SQLBindExec) dropSQLBind() error {
 	return err
 }
 
+func (e *SQLBindExec) dropSQLBindByDigest() error {
+	if !e.isGlobal {
```
Comment (qw4990) [inline executor/bind.go:92]: For safety, `e.sqlDigest` cannot be empty. Please returning an error here if `e.sqlDigest==""`.

File: bindinfo/session_handle.go [hunk 3]
```diff
@@ -98,11 +98,39 @@ func (h *SessionHandle) DropBindRecord(originalSQL, db string, binding *Binding)
...
+	return nil
+}
+
 // GetBindRecord return the BindMeta of the (normdOrigSQL,db) if BindMeta exist.
 func (h *SessionHandle) GetBindRecord(hash, normdOrigSQL, db string) *BindRecord {
 	return h.ch.GetBindRecord(hash, normdOrigSQL, db)
 }
 
+// GetBindRecordBySQLDigest return all BindMeta corresponding to hash.
+func (h *SessionHandle) GetBindRecordBySQLDigest(hash string) (*BindRecord, error) {
```
Comment (qw4990) [inline bindinfo/session_handle.go:130]: Better to use the unified name `sqlDigest` instead of `hash`.

File: bindinfo/handle.go [hunk 4]
```diff
@@ -423,6 +423,59 @@ func (h *BindHandle) DropBindRecord(originalSQL, db string, binding *Binding) (d
...
+
+	updateTs := types.NewTime(types.FromGoTime(time.Now()), mysql.TypeTimestamp, 3).String()
+
+	_, err = exec.ExecuteInternal(ctx, `UPDATE mysql.bind_info SET status = %?, update_time = %? WHERE original_sql = %? AND update_time < %? AND status != %?`,
+		deleted, updateTs, originalSQL, updateTs, deleted)
+	if err != nil {
+		return 0, err
+	}
+
+	return h.sctx.Context.GetSessionVars().StmtCtx.AffectedRows(), nil
```
Comment (qw4990) [inline bindinfo/handle.go:476]: Can we just use `h.DropBindRecord` here?

**Review by qw4990 — COMMENTED** [review]
And please add some UTs for `drop [global] binding for SQL {digeest}`.

File: bindinfo/bind_test.go [hunk 5]
```diff
@@ -1304,3 +1304,67 @@ func TestBindSQLDigest(t *testing.T) {
...
+		// storage hints
+		{"select * from t", "select /*+ read_from_storage(tikv[t]) */ * from t"},
+		// others
+		{"select t1.a, t1.b from t t1 where t1.a in (select t2.a from t t2)", "select /*+ use_toja(true) */ t1.a, t1.b from t t1 where t1.a in (select t2.a from t t2)"},
+	}
+
+	h := dom.BindHandle()
+	for _, c := range cases {
+		utilCleanBindingEnv(tk, dom)
+		sql := "create global binding for " + c.origin + " using " + c.hint
```
Comment (fzzf678) [inline bindinfo/bind_test.go:1357]: comment

File: planner/core/planbuilder.go [hunk 6]
```diff
@@ -979,14 +979,23 @@ func (b *PlanBuilder) buildSet(ctx context.Context, v *ast.SetStmt) (Plan, error
...
+		}
+		if v.HintedNode != nil {
+			p.BindSQL = utilparser.RestoreWithDefaultDB(v.HintedNode, b.ctx.GetSessionVars().CurrentDB, v.HintedNode.Text())
+		}
+	} else {
+		p = &SQLBindPlan{
+			SQLBindOp: OpSQLBindDropByDigest,
+			IsGlobal:  v.GlobalScope,
+			SQLDigest: v.SQLDigest,
+		}
```
Comment (fzzf678) [inline planner/core/planbuilder.go:998]: Should we make this `else{}` as a new function

File: bindinfo/handle.go [hunk 7]
```diff
@@ -658,6 +667,11 @@ func (h *BindHandle) GetBindRecord(hash, normdOrigSQL, db string) *BindRecord {
 	return h.bindInfo.Load().(*bindCache).GetBindRecord(hash, normdOrigSQL, db)
 }
 
+// GetBindRecordBySQLDigest returns the BindRecord of the sql digest.
+func (h *BindHandle) GetBindRecordBySQLDigest(hash string) (*BindRecord, error) {
```
Comment (qw4990) [inline bindinfo/handle.go:671]: `hash` --> `sqlDigest`

**Review by qw4990 — APPROVED** [review]

**Review by Reminiscent — APPROVED** [review]

**Pull request merged**